sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/UnsafeColumnVector.java

```diff
@@ -0,0 +1,517 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.vectorized;
+
+import java.nio.ByteBuffer;
+
+import org.apache.commons.lang.NotImplementedException;
+
+import org.apache.spark.sql.catalyst.expressions.UnsafeArrayData;
+import org.apache.spark.sql.types.*;
+import org.apache.spark.unsafe.Platform;
+
+/**
+ * A column backed by UnsafeArrayData on byte[].
+ */
+public final class UnsafeColumnVector extends WritableColumnVector {
```

Comment:
This abstraction looks confused at first glance. It seems not following some `ColumnVector` APIs usage. Looks like this uses `putByteArray` to set up byte array `data` which stores the data of this array column.

IIUC, this is proposed to represent only array column, but some APIs implementation looks weird. For example, `getBoolean` respects `rowId` parameter and `getBooleans` doesn't.

Reply:
You are right `UnsafeColumnVector.putByteArray` is used to put the whole array in `byte[]` for `UnsafeArrayData`. I will put some comment to explain the usage of this API to make it clear.

Good catch for `getBooleans`. It seems to be my fault since it has to take care of `rowId`. I will fix this.

Reply:
Since this `UnsafeColumnVector` represents array column, will we use the APIs like `getBoolean`, `getBooleans`...etc.?

Reply:
These two methods `getBoolean` and `getBooleans` are used [here](github.com/apache/spark/blob/dce1610ae376af00712ba7f4c99bfb4c006dbaec/sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarArray.java) and [there](github.com/apache/spark/blob/dce1610ae376af00712ba7f4c99bfb4c006dbaec/sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarArray.java) in `ColumnarArray`.